lib.fixedPoints.toExtension: init

ShamrockLee opened the pull request:
## Description of changes

This function is taken from `stdenv.mkDerivation`'s implementation of `overrideAttrs`, and acts as a `toFunction` analogy for overlays.

Some packages/package sets have an argument in their `override` interface to take override functions, which might be anything from a plain attribute set to an overlay. To use the override interface while keeping the previous overriding, one would extend it with `lib.composeExtensions`, which only accepts overlays.

#225051 locally uses a function like this to convert the `overrideModAttrs` argument to `passthru.overrideModAttrs`. The former could be anything to pass to `.overrideAttrs`, but the latter must be an attribute overlay so that people could override with `lib.composeExtension`.

Other potential use cases include enabling custom overriders such as `overridePythonAttrs` support overlay-style overriding (`drv.overridePythonAttrs (finalAttrs: previousPythonAttrs: { ... })`), but that is out of the scope of this PR.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux (no rebuild)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
lib.toExtension: init

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

ShamrockLee committed:
stdenv.mkDerivation: simplify overrideAttrs with extends and toOverlay

ShamrockLee committed:
lib/tests/misc.nix: move testFix into new category FIXED-POINTS

ShamrockLee commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>disko</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

roberth commented on lib/fixed-points.nix:
````diff
@@ -394,4 +394,64 @@ rec {
+    # Type
+
+    ```
+    toExtension ::: (a | a -> a | a -> a -> a) -> a -> a -> a
+    a = AttrSet & !Function
````
This is a bit more general and accurate.
- Implementation doesn't actually care about attribute sets, and would work equally well for lists, insofar as lists work well for anything.
- Make better use of parametric polymorphism. (`a` felt more like a "top" type, such as `Object` in Java or `unknown` in TypeScript)

```suggestion
    toExtension ::
      (data | b -> data | a -> b -> c) ->
        a -> b -> c
    where data = ! Callable
```

ShamrockLee replied:
Thank you for reviewing!

> ```nix
> toExtension ::
>   (data | b -> data | a -> b -> c) ->
>     a -> b -> c
> where data = ! Callable
> ```

Why are we using `c` instead of `data` here?

roberth replied:
Good point


> Why are we using `c` instead of `data` here?

- In the argument, in the latter `|` case, `c` could be anything, including a function.
- `c` can be returned by the returned extension function
  - but so can `data`!


```suggestion
    toExtension ::
      (data | b -> data | a -> b -> c) ->
        a -> b -> (data | c)
    where data = ! Callable
```

----

Do you think this is easier to read?

```suggestion
    toExtension ::
      ( data
      | b -> data
      | a -> b -> c
      ) ->
      a -> b -> (data | c)
    where data = ! Callable
```

ShamrockLee replied:
I think it's reasonable to expect `c = data`, as our goal is to construct a function in the form `final: prev: dataTemplate`, and non-function data will be returned directly.

Is there a counter-example in which `c != data`?

roberth replied:
There is. Here's an example that returns a function instead of attrs. I wouldn't recommend it, because the function needs to be invoked to be useful, which means these attributes aren't lazy like when you return them directly.

```nix
let
  extendsWith = op: overlay: f: final: let prev = f final; in op prev (overlay final prev)
in

fix
  (extendsWith
    (p: q: x: p x // q x)
    (toExtension (final: prev:
      x:
        let pre = prev x; fin = final x;
        in
          { a = 1; b = pre.a; c = fin.a + x; }))
    (final: x: { a = 0; c = final.a; }))
  42
```

roberth replied:
I now realize that `data` might be too suggestive. An attribute set _containing_ a function would be fine.

ShamrockLee replied:
The purpose of `toExtension` is to make `toExtension (final: prev: resultTemplate)`, `toExtension (prev: resultTemplate)`, and `toExtension result` all evaluate to `final: prev: resultTemplate`, which doesn't hold when the result is a function/callable.

Another implementation assumption is that `(final: prev: resultTemplate)` could take `prev` without error, which suggests that the "`prev`-space" is a subset of the "`final`-space." Therefore, the `prev` type is a sub-type of `final`. They are likely to share the same type in Nix Language.

ShamrockLee replied:
What do you think about this?

```suggestion
    toExtension ::
      (a | a -> b | a -> a -> b) -> a -> a -> b
    where a = ! Callable
```

For example:
```nix
toExtension { a = 1; b = 2; c = null; }
=> final: prev: { a = 1; b = 2; c = null; }
```
```nix
toExtension (prev: { a = 1; b = elemAt prev 0; c = null; })
=> final: prev: { a = 1; b = elemAt prev 0;  c = null; }
```
```nix
toExtension (final: prev: { a = 1; b = elemAt prev 0; c = (elemAt final 0) + 1; })
=> final: prev: { a = 1; b = elemAt prev 0;  c = (elemAt final 0) + 1; }
```

We could replicate the documented examples with custom `fix'` and `extends'`:

```nix
elemAt = lib.elemAt

findFirst = lib.findFirst

toExtension = lib.toExtension

zipAttrsWith = lib.zipAttrsWith

recover = x: [ x.a x.b x.c ]

update' = lSet: rSet: zipAttrsWith (name: findFirst (x: x != null) null) [ rSet lSet ]

fix' = f: let x = f (recover x); in x

extends' = g: f: final: update' (f final) (g final (recover (f final)))

fix' (final: { a = 0; b = null; c = elemAt final 0; })
=> { a = 0; b = null; c = 0 }

fix' (extends' (toExtension { a = 1; b = 2; c = null; }) (final: { a = 0; b = null; c = elemAt final 0; }))
=> { a = 1; b = 2; c = 1; }

fix' (extends' (toExtension (prev: { a = 1; b = elemAt prev 0; c = null; })) (final: { a = 0; b = null; c = (elemAt final 0) + 1; }))
=> { a = 1; b = 0; c = 2; }
```

roberth replied:
> The purpose of `toExtension` is to make `toExtension (final: prev: resultTemplate)`, `toExtension (prev: resultTemplate)`, and `toExtension result` all evaluate to `final: prev: resultTemplate`, which doesn't hold when the result is a function/callable.

This means that `toExtension` is subjectively not a great fit for such a use case, but it doesn't mean that it is incapable of producing a useful result. It can still be used for the only input type that isn't "misclassified", `a -> a -> b`.
It does not mean that the type of the function is less general than it really is.

I think when writing types, we should try to err on the side of being more general, so that we don't mislead users into thinking that a function is not applicable to their use case. There's no way to find or correct that mistake, so I'd rather have them try and fail. That might sound bad but I don't think it's a big deal, because I expect such false positives to be quite rare, especially because we also have documentation for them to judge the function's fitness for their purpose.


> Another implementation assumption is that `(final: prev: resultTemplate)` could take `prev` without error, which suggests that the "`prev`-space" is a subset of the "`final`-space." Therefore, the `prev` type is a sub-type of `final`. They are likely to share the same type in Nix Language.

This is not an intrinsic property of the function, so I don't think it belongs in the type.
For instance, the `extendsWith` function I showed would allow you to merge extensions however you want, changing the next layer (and therefore `final`) to have a wildly different structure, possibly.

ShamrockLee replied:
Considering the current implementation, the most general input and output that won't cause type error would be

```suggestion
    toExtension ::
      b' -> Any -> Any -> b'

    or

    toExtension ::
      (a -> b') -> Any -> a -> b'

    or

    toExtension ::
      (a -> a -> b') -> a -> a -> b'

    where b' = ! Callable

    If a = b = b' = AttrSet & ! Callable, toExtension returns an extending function.
```

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,64 @@ rec {
+  toExtension = f0:
+  if lib.isFunction f0 then
+    final: prev:
+      if lib.isFunction (f0 prev) then
+        # f0 is (final: prev: { ... })
+        f0 final prev
+      else
+        # f0 is (prev: { ... })
+        f0 prev
```
```suggestion
      let fPrev = f0 prev;
      if lib.isFunction (f0 prev) then
        # f0 is (final: prev: { ... })
        f0 final prev
      else
        # f0 is (prev: { ... })
        fPrev
```

`f0 prev` may (or may not) be expensive to recompute.
Let's be on the safe side.

I think you could rename `f0` to `f` as well, because there's no ambiguity with other `f`s. Or even `v`, because we don't really know whether `f` is even a function, as in the final case.

roberth commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -67,26 +67,16 @@ let
-        in
-          makeDerivationExtensible
-            (self: let super = rattrs self; in super // (if builtins.isFunction f0 || f0?__functor then f self super else f0));
+        # Convert f0 to an overlay and use it to extend rattrs.
+        # Legacy is:
+        #   overrideAttrs (super: {})
+        # We want to introduce self. We follow the convention of overlays:
+        #   overrideAttrs (self: super: {})
+        # Which means the first parameter can be either self or super.
+        # This is surprising, but far better than the confusion that would
+        # arise from flipping an overlay's parameters in some cases.
```
```suggestion
```

These comments of mine aren't relevant anymore.

roberth reviewed (commented):
Great idea!

roberth committed:
lib.fixedPoints.toExtension: Improve type

This makes the type more informative and general - closer to how the function behaves, particularly in isolation.

Co-authored-by: Yueh-Shun Li <shamrocklee@posteo.net>

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+
+    `toExtension` is the `toFunction` for extending functions (a.k.a. extensions or overlays).
+    It converts a non-function or a single-argument function to an extending function,
+    while returning a double-argument function as-is.
```
```suggestion
    while returning a two-argument function as-is.
```

Maybe it's just me, but double sounds a bit like an intrinsic property, which it isn't here.

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+    It converts a non-function or a single-argument function to an extending function,
+    while returning a double-argument function as-is.
+
+    That is, it takes one of `x`, `prev: x`, or `final: prev: x`,
+    and returns `final: prev: x`, where `x` is not a function.
```
```suggestion
    That is, it takes a value of the shape `x`, `prev: x`, or `final: prev: x`,
    and returns `final: prev: x`, assuming `x` is not a function.
```

- value: We're not thinking about expressions.
- where -> assuming: It's not a requirement on `toExpression x`, but a precondition `x ->` the preceding statement.

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+    That is, it takes one of `x`, `prev: x`, or `final: prev: x`,
+    and returns `final: prev: x`, where `x` is not a function.
+
+    This function is extracted from the implementation of
+    the fixed-point arguments support of `stdenv.mkDerivation`.
```
```suggestion
    This function takes care of the input to `stdenv.mkDerivation`'s
    `overrideAttrs` function.
```

Describe the code, not the change.

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+    This function is extracted from the implementation of
+    the fixed-point arguments support of `stdenv.mkDerivation`.
+    It bridges the gap between `<pkg>.overrideAttrs`
+    before and after the overlay-style support,
+    as well as `config.packageOverrides` and `config.overlays` in `pkgs`.
```
```suggestion
    before and after the overlay-style support.
```

I think this goes into too much detail.
Also I don't think `config.overlays` uses it.

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+    # Inputs
+
+    `f`
+    : The function or non-function to convert to an extending function.
```
```suggestion
    : The function or value to convert to an extending function.
```

This looks a little less awkward IMO.

It could be further simplified to "the value", but then we're losing a good hint that it's often a function, so I think "function or value" is best, despite functions being values.

roberth commented on lib/fixed-points.nix:
```diff
@@ -394,4 +394,77 @@ rec {
+        # f is (prev: { ... })
+        fPrev
+    else
+      # f is { ... }
```
```suggestion
      # f is not a function; probably { ... }
```

roberth reviewed (commented):
Have a few suggestions for the docs, but otherwise LGTM!

ShamrockLee committed:
lib.fixedPoints.toExtension: improve documentation

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

ShamrockLee committed:
buildGoModule: use lib.toExtension

Use toExtension from Nixpkgs library instead of the one defined locally.

doronbehar commented:
Thanks for this PR, it is a little bit above my league and above my current level of available focus to review it properly, but it looks good and useful. Do you think it'd be possible to use these changes to help with https://github.com/NixOS/nixpkgs/issues/44426 ?

roberth commented:
Thank you @ShamrockLee!

@doronbehar Yes, it seems like a useful function for that, especially if it needs to support both the arity-1 and arity-2 style overriding like `overrideAttrs` does. After calling this, you always have overlays, and you can work with the functions for overlays to handle them.
